ext/app/electron/DocRegistry.ts

```diff
@@ -0,0 +1,74 @@
+import * as path from "path";
+import { HomeDBManager } from "app/gen-server/lib/homedb/HomeDBManager";
+import { fileExists } from "./utils";
+
+export class DocRegistry {
+
+  private idToPathMap: Map<string, string>;
+  private pathToIdMap: Map<string, string>;
+  private db: HomeDBManager;
+
```

Comment:
I have the same concerns! DocRegistry is a not-so-great workaround in the context of a larger problem.

The core logic that powers this PR is the `getPath` override. Note that `getPath` used to be synchronous, while a TypeORM database lookup is async, so the new `getPath` cannot invoke a DB lookup unless we make `getPath` async. Updating it to be async turned out to be a bottomless rabbit hole involving tons of other functions that must also be made async. @paulfitz suggested making all `externalId` info in-memory, so I created the DocRegistry class to hold an in-memory cache.

I tried my best to avoid cache inconsistency:

- Grist Desktop has a single instance lock, so if any cached info is updated, it must be updated by the only running instance which should know about the update and keep the cache coherent.
- There's nothing I could do to prevent them from doing so, but users are not supposed to edit the database manually.
- A document being moved or deleted on the FS level can leave invalid entries in the DB. It will be very involved to watch for such FS updates (possible via `inotify` on POSIX systems, not sure about Windows) and take action accordingly. Desktop apps usually throw an error like "file does not exist" when you access a stale cache record, then delete the cache entry, but for Grist this is not straightforward to implement due to the way Grist creates new documents. Note that in this case the cache is still consistent with the DB, and there's a toy GC mechanism to remove such stale entries each time Grist Desktop starts.

Reply:
This is amazing context again Leslie, really appreciating the detailed writeups!

I'll have a think about whether there's another workaround that doesn't involve caching.

Or maybe at least, involves caching closer to Grist Core / Storage management, to reduce the chance of inconsistencies popping up!

It might be that the work I'm doing to enable replacing DocStorageManager, might enable another solution here.

Reply:
I looked into this and couldn't find a better approach without substantial work. 

I've merged DocRegistry into the new DesktopDocStorageManager class to get it a little bit closer to where the actual paths are handled, and hopefully an opportunity opens up in the future to figure this out.